cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs

```diff
@@ -164,12 +164,12 @@ type VROrigin = (VR, ExposedComponent, DependencyReason QPN)
 type Conflict = (ConflictSet, FailReason)
 
 validate :: Tree d c -> Validate (Tree d c)
-validate = cata go
+validate = go
   where
-    go :: TreeF d c (Validate (Tree d c)) -> Validate (Tree d c)
+    go :: Tree d c -> Validate (Tree d c)
 
-    go (PChoiceF qpn rdm gr       ts) = PChoice qpn rdm gr <$> sequence (W.mapWithKey (goP qpn) ts)
-    go (FChoiceF qfn rdm gr b m d ts) =
+    go (PChoice qpn rdm gr       ts) = PChoice qpn rdm gr <$> sequence (W.mapWithKey (goP qpn) (fmap go ts))
```

Comment:
sequence . map seems bad idea. Should we add `traverseWithKey` to `WeightedPSQ` (separately)? Maybe it fuses, but  I'm somewhat sceptical about that.

Reply:
Sure. I take it you don't consider that a blocker for this PR, as its about changing a different thing than this PR currently does. Are you ok with approving this pr as is?

Reply:
I can fix it up with `traverseWithKey` in the follow up #7533

Reply:
Added it first, then pushed to this branch.